design-proposals/klip-49-add-source-stream-table-semantic.md

```diff
@@ -0,0 +1,129 @@
+part of the computation resource. This would be a special type of running query without `SELECT` clause or a query id, as it 
+should be only associated with source table, and user could choose to `DROP` the entire table if they don't want to use resource 
+to do the materialization.
+
+Here comes a precaution we might be aware. The original `CREATE TABLE` command does not need materialization, so there is 
+no enforcement for primary key. However, in order to let pull query run on the materialized state, the store must be keyed so that 
+pull query could use the primary key for access.  
+
+To enforce primary key on the `CREATE SOURCE TABLE` command, we have 3 options:
+1. `AstBuilder` could verify that at least one `TableElement` contains primary key, otherwise the plan fails.
+2. `LogicalPlanner#buildPersistentLogicalPlan` could verify the plan node and see whether it needs materialization 
+but has no defined primary key.
+3. Create an injector similar to `TopicCreateInjector`, which could be used to validate things through `RequestValidator`
+
+In long-term, we don't want to bloat `AstBuilder` with all the logical constraint checks. In the short-term, the complexity 
+of #1 is low, and the expected code change to enforce this constraint should be low, so we would pick #1. 
+
+## Public APIS
+* Create source table/stream, which makes the data source read-only:
+
+```roomsql
+CREATE SOURCE STREAM stream_name (COL1 INT, COL2 STRING) WITH ...
+CREATE SOURCE TABLE table_name (ID INT PRIMARY KEY, COL1 INT) WITH ...
+```
+
+## Test plan
+The new semantics will be tested as:
+
+* Unit testing for modular changes
+* QTTs with the new source table/stream syntax, including insertion failure and successful materialization
+* Tests for `DESCRIBE` sources
+
+## LOEs and Delivery Milestones
+The estimated work time will be spent as follows (to implement and review):
+1. Add `SOURCE` syntax for both stream/table takes a week
+2. Make changes in Metastore and CreateSource for read-only sources takes a week to review
+3. Reject insertion to `SOURCE` table/stream takes a week
+4. Display read-only attributes on `DESCRIBE ... EXTENDED` 3 days 
+5. Source table materialization takes a week
+6. Add pull query access for source table takes about 3 days
+7. Documentation changes takes 3 days
+8. Integration testing takes 2 weeks
+
+There will be 3 milestones, first is the completion of syntax addition (#1), which marks the start 
+point of splitting the work between engineers. Second milestone will be the feature complete (#2~6), and 
+the last one will be documentation and tests complete.
+
+## Documentation Updates
+* Add a new doc page as `docs/developer-guide/ksqldb-reference/source-table.md`
+* Add a new doc page as `docs/developer-guide/ksqldb-reference/source-stream.md`
+
+## Compatibility Implications
+This is a new feature in CT and CS, which has no backward compatibility issue. 
+
+## Security Implications
+None
```

Comment:
I think it might make sense to explicitly call out that this is _not_ a security feature. For the scope of this PR it seems that you would still be allowed to `CREATE TABLE` over an existing topic (not `SOURCE TABLE`) and that we're not deprecating that functionality yet. Is that correct?

Reply:
Not sure I follow you here, are you asking whether it is allowed to have multiple source/non-source table being created on top of the same topic?

Reply:
yup, that's essentially what I meant

Reply:
That's a good question, IMHO, if we eventually have a full RBAC-like model for KSQL, it is logistic to create multiple tables on the same topic with different user roles, where the `admin` could create the table with write access, while a `consumer` role should be only allowed to create read-only table. I could mention explicitly here that we don't want to enforce such restriction at the moment for no obvious benefit.